apps/frontend/src/app/Solver/GOSolver/linearUB.ts

```diff
@@ -0,0 +1,80 @@
+import { OptNode } from "../../Formula/optimization";
+import { cartesian } from "../../Util/Util";
+import { ArtifactsBySlot, MinMax, computeFullArtRange } from "../common";
+import { Linear } from "./BNBSplitWorker";
+import { polyUB } from "./polyUB";
+import { solveLP } from "./solveLP_simplex";
+
+function weightedSum(...entries: readonly (readonly [number, Linear])[]): Linear {
+  const result = { $c: 0 }
+  for (const [weight, entry] of entries)
+    for (const [k, v] of Object.entries(entry))
+      result[k] = (result[k] ?? 0) + weight * v
+  return result
+}
+
+export function linearUB(nodes: OptNode[], arts: ArtifactsBySlot) {
```

Comment:
This is exposed API, please add return type. If it is complex, make a new `type` declaration.